### ISSUE TITLE
`Parser` may be incorrectly named

### ISSUE THREAD

**Issue by seagreen** [bos/aeson#615]
I've been confused about this for a while, but didn't realize what was going on until I saw it discussed [in this tutorial](https://artyom.me/aeson):

> Aeson has a type called `Parser`, but it's not a type for parsers – it's a type for results of parsers.
...
> So, all parsers actually have the type `Value -> Parser a`.

This is a fairly serious UI issue for the library, so I think it's worth discussing.

As far as concrete steps go to improve the situation, one would be to improve the Haddocks for `Parser`. Currently [they're](https://github.com/bos/aeson/blob/22488951884c4a17f1d470324822f8c3ec09d719/Data/Aeson/Types/Internal.hs#L273):
```
-- | A JSON parser.
```
If the linked tutorial is correct, it would be nice to clarify that this isn't actually a parser but a result type. It would also be nice to say what the type of Aeson parsers actually are -- the tutorial says they're `Value -> Parser a`s, but I think they may be more general than that.

At this point the user will probably be wondering (I have this question myself) how this differs from the other `Result` type, especially since they show up in the same type signatures, eg `parse :: (a -> Parser b) -> a -> Result b`. So it would be nice to document that as well.

**Comment by Lysxia** [comment]
I wonder how much it would break to rename `Parser` to `Result` or something, but keep a `Parser` synonym around.

**Comment by phadej** [comment]
FWIW, `aeson` has `Result` as well ;)

**Comment by bergmark** [comment]
> Aeson has a type called Parser, but it's not a type for parsers – it's a type for results of parsers.

I'm not sure I agree with this; Doesn't the result of parsing depend on how you invoke the parser? e.g. using `eitherDecode` the result is `Either String a`. And as @phadej notes we have the `Result` type as well, you normally don't see it as a user but it's currently isomorphic to `Either String a`.

`Parser a`  defines a parser for `a`. I'd consider the fact that you need to go through `Value` to get your `a` an implementation detail. We have in fact discussed removing this requirement as we could allow more flexible and efficient parsing without it.

So we can split these into three different concepts in aeson which from what I've seen is the same as in other libraries such as parsec: Parsers (`Parser a`, `ParsecT s u m a`), parse results (`Either String a`, `Result a`), and runners for parsers (`eitherDecode :: ByteString -> Either String a`, ``, et.c.).

**Comment by seagreen** [comment]
> I'm not sure I agree with this

I'm not sure I do either:) Let me lay out my thinking though so we have a common ground to start from.

The way I've always seen parsers defined is some variation of this:
```
newtype FooParser :: Input -> Maybe (Foo, Input)
```

To pick a concrete example we can look at the [Write You a Haskell toy parser](http://dev.stephendiehl.com/fun/002_parsers.html):
```
newtype Parser a = Parser { parse :: String -> [(a,String)] }
```

The `attoparsec` [Parser](https://hackage.haskell.org/package/attoparsec-0.13.2.0/docs/Data-Attoparsec-Internal-Types.html#t:Parser) is more complicated, but you can see that the current state/input is within the newtype. So `attoparsec`'s [parse functions](https://hackage.haskell.org/package/attoparsec-0.13.2.0/docs/Data-Attoparsec-Text.html#v:parse) look like:
```
parse :: Parser a -> Text -> Result a
```
Instead of `aeson`'s [which is](https://hackage.haskell.org/package/aeson-1.2.3.0/docs/Data-Aeson-Types.html#v:parse):
```
parse :: (a -> Parser b) -> a -> Result b
```
If `aeson`'s Parser was of the form of the other examples, its `parse` function would be:
```
parse :: Parser b -> a -> Result b
```

To address the other issue, happily I don't think using the `attoparsec` strategy means committing to a single concrete input type. `attoparsec`'s Parser is parameterized over the input.

**Comment by phadej** [comment]
Aeson's parser have to be `Parser a b`, we have different input types. E.g. type of commonly used `withObject` will be `withObject :: String -> Parser Object b -> Parser Value b`. Note, this `Parser` is a `Profunctor`, special case of `Star`: `Parser a b ~ Star CurrentParser a b`.

Not hidden (in a newtype) `->` makes writing special cases parses easy. I challenge Ian: define a `newtype Pro a b = Pro { runPro :: a -> Parser a b }`, and using this newtype write a `Pro Value Int` which would accept `Null` as 0 and `Bool` as 0&1 in addition to `Number`. I think you have to define bunch of helpers (like takeNull). That API might turn out nice, but it's not a small task.

**Comment by tomjaguarpaw** [comment]
At the very least `Parser` needs an unmissable comment explaining that it's not what we typically think of as a parser.  It's a parser to which the input has already been applied!

**Comment by tomjaguarpaw** [comment]
My first attempt at improving the situation.  Comments gratefully received:

https://github.com/bos/aeson/pull/635

**Comment by bergmark** [comment]
Resolved by #635
